### PR TITLE
Standard router

### PR DESCRIPTION
**Pull request by jsnoble**
- introduction of a general dynamic router that uses apis to route data
- currently only works with elasticsearch

### CODE REVIEW

**Comment by jsnoble** [comment]
// Example Job
{
    "name": "Multi Routes",
    "lifecycle": "once",
    "analytics": true,
    "slicers": 1,
    "workers": 1,
    "assets": [
        "elasticsearch",
        "standard"
    ],
    "apis": [
        {
            "_name": "elasticsearch_sender_api"
        }
    ],
    "operations": [
        {
            "_op": "elasticsearch_reader",
            "index": "test_index",
            "size": 5000,
            "date_field_name": "created"
          },
          {
            "_op": "elasticsearch_index_selector",
            "index": "multi_route",
            "preserve_id": true,
            "type": "events"
          },
          {
            "_op": "key_router",
            "use": 1,
            "from": "beginning"
          },
          {
            "_op": "routed_sender",
            "api_name": "elasticsearch_sender_api",
            "routing" : {
                "A,a" : "es7",
                "*": "default"
            },
            "size": 10
          }
    ]
}

File: asset/src/routed_sender/schema.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+export default class Schema extends ConvictSchema<RouteSenderConfig> {
+    build(): AnyObject {
+        return {
+            size: {
+                doc: 'the maximum number of docs it will take at a time, anything past it will be split up and sent'
+                + 'note that the value should be even, the first doc will be the index data and then the next is the data',
+                default: 500,
+                format(val: any) {
+                    if (isNaN(val)) {
+                        throw new Error('Invalid size parameter for elasticsearch_bulk opConfig, it must be a number');
```
Comment (peterdemartini) [inline asset/src/routed_sender/schema.ts:13]: This references the `elasticsearch_bulk`, be careful about copy and paste

File: asset/src/routed_sender/processor.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+                const routeConfig = this.routingExectuion.get('**') as Endpoint;
+                routeConfig.data.push(record);
+            } else {
+                let error: TSError;
+
+                if (route == null) {
+                    error = new TSError('No route was specified in record metadata');
+                } else {
+                    error = new TSError(`Invalid connection route: ${route} was not found on connector_map`);
```
Comment (peterdemartini) [inline asset/src/routed_sender/processor.ts:91]:
```suggestion
                    error = new TSError(`Invalid connection route: ${route} was not found in routing`);
```

File: asset/src/routed_sender/processor.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+        const senders = [];
+
+        for (const [, { data, client }] of this.routingExectuion) {
+            if (data.length > 0) {
+                senders.push(...this.sendRecords(client, data));
+            }
+        }
+
+        await Promise.all(senders);
```
Comment (peterdemartini) [inline asset/src/routed_sender/processor.ts:106]: This should probably use `pMap` with some sort of concurrency control or maybe it should run serially?

**Pull request merged**